ls1intum/Artemis
# Add transformation modeling exercises

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


This change introduces transformation modeling exercises, an extension to the already implemented modeling exercises, where the problem statement can include an additional Apollon diagram. Students are asked to transform this given diagram into another type of diagram, for instance, computing the reachability graph of a given Petri net. Submissions are designed to be automatically assessed as soon they are uploaded to the server. The grading is done based on a predefined scheme by the instructor.

### Description

A new exercise type was added that extends the existing modeling exercise. It includes new properties for the problem statement diagram and a grading scheme that defines how the solution should be graded. At the moment, the transformation from Petri nets to reachability graphs is supported, however, additional transformations may be implemented. For this transformation, submissions are graded using the coverability graph algorithm to decide whether the students' submission matches the precomputed correct solution. This is achieved by a separate service. 

### Steps for Testing

Prerequisites:
- 1 instructor
- 1 student
- 1 transformation modeling exercise in a course

1. Create a transformation modeling exercise
2. Create a submission for this exercise
3. Check whether it was graded by the server and if the feedback was added to the submissions' result

### Review Progress





#### Code Review
#### Manual Tests

### Screenshots


Screenshots of the various features with extensive explanation of the functionality: https://home.in.tum.de/~dejori/ba.pdf
